### ISSUE TITLE
Default annotation value not propagated with @AliasFor

### ISSUE THREAD

**Issue by l0co** [micronaut-projects/micronaut-core#5036]
In `2.3.4`.

Looking on these classes:

```java
@Retryable
public @interface CircuitBreaker {
    @AliasFor(annotation = Retryable.class, member = "delay")
    String delay() default "500ms";
}

@Around
@Type(DefaultRetryInterceptor.class)
public @interface Retryable {
    String delay() default "1s";
}
```

It seems that when I use `@CircuitBreak` with no params the delay should be 500ms, while, because it's a default value, it's not propagated by `@AliasFor` to `@Retryable` and finally the delay calculated by `DefaultRetryInterceptor` is 1s.

This only works when I explicitely use `@CircuitBreak(delay = "500ms")`.

**Comment by jameskleeh** [comment]
I don't think we can change this because it would be impossible to know if the user set the value or not. Perhaps we could have some option in the `@AliasFor` annotation to allow it

**Comment by l0co** [comment]
Why do you need to know if user set the value or we use the default value? If user set the value on their annotation it should be propagated upward annotations hierarchy with `@AliasFor`. If user hasn't set the value, the default annotation value should be propagated. Otherwise the current annotation contract is broken.

Currently, you cannot even have the annotation with other defaults than in original annotations and it would be nice to have this feature in Micronaut. Consider this example:

```java
@Client
@CircuitBreaker
public @interface SafeClient {

	@AliasFor(annotation = Client.class, member = "value")
	String serviceId();

	@AliasFor(annotation = Client.class, member = "path")
	String servicePath();

	@AliasFor(annotation = Client.class, member = "configuration")
	Class<? extends HttpClientConfiguration> configuration() default HttpClientConfiguration.class;

	@AliasFor(annotation = CircuitBreaker.class, member = "includes")
	Class<? extends Throwable>[] includes() /*default { ServiceUnavailableException.class }*/;

	@Digits(integer = CircuitBreaker.MAX_RETRY_ATTEMPTS, fraction = 0)
	@AliasFor(annotation = CircuitBreaker.class, member = "attempts")
	String attempts() default "5";

	@AliasFor(annotation = CircuitBreaker.class, member = "delay")
	String delay() default "100ms";

	@AliasFor(annotation = CircuitBreaker.class, member = "reset")
	String reset() default "10s";

}
```

This annotation is supposed to add automatically `@Client` and `@CircuitBreaker` annotations to some interface, with some default configuration. But currently it's pointless because default values aren't propagated to `@Client` and `@CircuitBreaker` annots.

**Comment by jameskleeh** [comment]
> Why do you need to know if user set the value or we use the default value

It's important in many areas. For example `metadata.stringValue()` which returns an optional would never return an empty optional if the default value is an empty string. There can be logic that changes behavior depending on the value being set by the user or not.

I understand that you want it to work that way, however we can't simply change the behavior. As I said one option is to have a member on AliasFor to change the behavior for that instance

**Comment by l0co** [comment]
Another thing here is:

```java
@CircuitBreaker
public @interface SafeClient {

	@AliasFor(annotation = Retryable.class, member = "attempts")
	String attempts() default "3";

}

@Retryable
public @interface CircuitBreaker {

    @AliasFor(annotation = Retryable.class, member = "attempts")
    String attempts() default "3";

}
```

After using `@SafeClient(attempts = "10")` what would be the result of `context.getAnnotation(Retryable.class).get("attempts", String.class)`? AFAIR when I was debugging this, it's still "3", what means `@AliasFor` is not propagated in a chain, but it's only propagated to the nearest annotation (which in this scenario is `@CircutBreaker`).

It was yesterday I was debugging it so I'm not 100% sure of this. But, if the improvement will be implemented I'd look at this as well.

**Comment by l0co** [comment]
> I understand that you want it to work that way, however we can't simply change the behavior. As I said one option is to have a member on AliasFor to change the behavior for that instance

I think it's fine to have this as an option.

**Comment by graemerocher** [comment]
Triaged as Feature/enhancement and closed completed. Current @AliasFor exposes applyDefault (since 5.2), and the annotation-metadata builder implements OverridesAttribute-style default propagation; merged #12911 includes coverage. This provides the requested opt-in behavior while retaining explicit-only alias semantics by default.
